### PR TITLE
[pytorch/profiler] Profiler NCCL metadata can now contain collective Input and Ouput Tensor addrs

### PR DESCRIPTION
**Pull request by sanrise**
Summary: Studying memory access patterns is the primary use cases.

Differential Revision: D65918359

### CODE REVIEW

**Comment by sanrise** [comment]
https://github.com/pytorch/pytorch/pull/139837 was reverted since there was a LOG line. I also took this chance to fix another suppressed issue with reading record function inputs in start callback and outputs in exit callback. 

earlier we were trying to read inputs in the exit callback and that would crash when running code without any sanitizers (`nosan`), this should no longer happen.

File: torch/csrc/profiler/util.cpp [hunk 1]
```diff
@@ -788,4 +909,54 @@ uint64_t computeFlops(
...
+    const char* fn_name) {
+  auto num_inputs = fn.num_inputs();
+  const auto inputs = fn.inputs();
+  VLOG(2) << "inputs: " << num_inputs << " " << inputs.size() << '\n';
+  // We have two cases: for unboxed kernel, we have num_inputs ==
+  // inputs.size() for boxed kernel using stack, there could be more elements
+  // on the stack from previous ops.
+  // TORCH_INTERNAL_ASSERT(num_inputs <= inputs.size());
+  if (num_inputs > inputs.size()) {
+    // LOG(WARNING) << "RecordFunction " << fn_name
```
Comment (briancoutinho) [inline torch/csrc/profiler/util.cpp:954]: remove?

**Review by briancoutinho — APPROVED** [review]
lgtm :)

File: torch/csrc/profiler/util.cpp (hunk 1 above)
Comment (sanrise) [reply torch/csrc/profiler/util.cpp:954]: I can remove but I also did not want to lost information on what it means when num_inputs > inputs.size(). Maybe I will just convert it to a comment

**Pull request closed without merging**